armi/reactor/converters/axialExpansionChanger/axialExpansionChanger.py

```diff
@@ -413,17 +484,269 @@ def manageCoreMesh(self, r):
+
+                    Instead, a mass weighted average temperature of {newToCompTemp} will be used. The consequence is that
+                    mass conservation is no longer guaranteed for this component type on this assembly!
+                    """  # noqa: E501
+                    runLog.warning(dedent(msg), label="Temp Search Failure")
+            except Exception as ee:
+                raise ee
+
+        # Do not use component.setTemperature as this mucks with the number densities we just calculated.
+        self.toComp.temperatureInC = newToCompTemp
+        self.toComp.clearCache()
+
+    def updateBOLParams(self):
+        """Update the BOL molesHmBOL and massHmBOL to stay consistent with the change in mass.
+
+        Notes
+        -----
+        Unlike the non-BOL information determined in py:meth:`setNewToCompNDens` and :py:meth:`setNewToCompTemperature`,
+        a new BOL temperature is not neccessary to be calculated here since we do not compute the ``massHmBOL`` or
+        ``moledHmBOL`` information on-the-fly; this allows us to directly set the new values. However, we do need to
+        compute the number densities and manually shift mass to account for the possibility of varying temperatures
+        between toComp and fromComp at BOL.
+        """
+        toCompVolBOL = self.toComp.getArea(Tc=self.toComp.p.temperatureInCBOL) * self.toComp.parent.p.heightBOL
+        fromCompVolBOL = self.fromComp.getArea(Tc=self.fromComp.p.temperatureInCBOL) * abs(self.deltaZTop)
+        newBOLVol = toCompVolBOL + fromCompVolBOL
+        newNDensBOL: dict[str, float] = {}
+        fromCompNDensBOL = dict(zip(self.fromCompBOLNucs, self.fromComp.p.numberDensitiesBOL))
+        toCompNDensBOL = dict(zip(self.toCompBOLNucs, self.toComp.p.numberDensitiesBOL))
+        nucsBOL = self._getAllNucs(self.toCompBOLNucs, self.fromCompBOLNucs)
+        for nuc in filter(nucDir.isHeavyMetal, nucsBOL):
+            massByNucFromCompBOL = densityTools.getMassInGrams(nuc, fromCompVolBOL, fromCompNDensBOL.get(nuc, 0.0))
+            massByNucToCompBOL = densityTools.getMassInGrams(nuc, toCompVolBOL, toCompNDensBOL.get(nuc, 0.0))
+            newNDensBOL[nuc] = densityTools.calculateNumberDensity(
+                nuc, massByNucFromCompBOL + massByNucToCompBOL, newBOLVol
+            )
+        self.toComp.p.molesHmBOL = sum(list(newNDensBOL.values())) / units.MOLES_PER_CC_TO_ATOMS_PER_BARN_CM * newBOLVol
+        self.toComp.p.massHmBOL = densityTools.calculateMassDensity(newNDensBOL) * newBOLVol
+
+        # update BOL Params for fromComp
+        self.fromComp.p.molesHmBOL *= 1.0 - (abs(self.deltaZTop) / self.fromComp.parent.p.heightBOL)
+        self.fromComp.p.massHmBOL *= 1.0 - (abs(self.deltaZTop) / self.fromComp.parent.p.heightBOL)
+
+    @staticmethod
+    def _sortKey(item):
+        match = re.search(r"([a-zA-Z]{1,2})(\d{1,3})?([a-zA-Z])?", item)
```

Comment:
We need a comment here.

No one glancing at this code is going to be able to quickly-and-easily read this. Our goal in writing code, especially in Python, is to make it easy for other people to read later.

For instance, I have no idea what this says, or what edge cases you are catching.